File: backend/app/utils/health_calculations.py
from __future__ import annotations

from typing import Any


def clamp(value: float, min_value: float, max_value: float) -> float:
    return min(max_value, max(min_value, value))


def calc_bmi(weight_kg: float, height_cm: float) -> float:
    height_m = float(height_cm) / 100
    if not height_m or not weight_kg:
        return 0.0
    return round(float(weight_kg) / (height_m * height_m), 1)


def normalize_sugar(value: str | None) -> str:
    if not value:
        return "Moderate"
    lowered = value.strip().lower()
    if lowered in {"low", "high", "moderate"}:
        return lowered.title()
    return value


def risk_level_from_percent(percent: float) -> str:
    if percent < 20:
        return "Low"
    if percent < 40:
        return "Moderate"
    return "High"


def profile_to_dict(profile: Any) -> dict[str, Any]:
    if isinstance(profile, dict):
        data = dict(profile)
    else:
        data = {
            "age": profile.age,
            "gender": profile.gender,
            "height_cm": profile.height_cm,
            "weight_kg": profile.weight_kg,
            "daily_steps": profile.daily_steps,
            "exercise_frequency": profile.exercise_frequency,
            "exercise_minutes": getattr(profile, "exercise_minutes", 25),
            "sleep_hours": profile.sleep_hours,
            "water_intake_liters": profile.water_intake_liters,
            "smoking": profile.smoking,
            "alcohol_consumption": profile.alcohol_consumption,
            "systolic_bp": profile.systolic_bp,
            "diastolic_bp": profile.diastolic_bp,
            "resting_heart_rate": profile.resting_heart_rate,
            "family_history": list(profile.family_history or []),
            "existing_conditions": list(profile.existing_conditions or []),
            "fruit_vegetable_intake": profile.fruit_vegetable_intake,
            "processed_food_frequency": profile.processed_food_frequency,
            "sugar_intake": profile.sugar_intake,
        }
    data["sugar_intake"] = normalize_sugar(data.get("sugar_intake"))
    data["family_history"] = list(data.get("family_history") or [])
    data["exercise_minutes"] = int(data.get("exercise_minutes") or 25)
    return data


def calculate_health_score(profile: Any) -> int:
    data = profile_to_dict(profile)
    bmi = calc_bmi(data["weight_kg"], data["height_cm"])
    score = 78.0

    if 18.5 <= bmi < 25:
        score += 6
    elif 25 <= bmi < 30:
        score -= 6
    elif bmi >= 30:
        score -= 12
    else:
        score -= 4

    score += clamp((data["daily_steps"] - 6000) / 400, -8, 8)
    score += clamp((data["sleep_hours"] - 7) * 4, -8, 6)
    score += clamp((data["exercise_minutes"] - 20) / 8, -6, 6)
    score += clamp((data["water_intake_liters"] - 2) * 3, -5, 4)

    sugar = data["sugar_intake"]
    if sugar == "High":
        score -= 7
    elif sugar == "Moderate":
        score -= 3
    else:
        score += 2

    smoking = str(data.get("smoking") or "")
    if smoking == "Current":
        score -= 10
    elif smoking == "Former":
        score -= 3

    score -= len(data["family_history"]) * 2
    return int(round(clamp(score, 28, 96)))


def calculate_overall_risk(profile: Any) -> int:
    data = profile_to_dict(profile)
    bmi = calc_bmi(data["weight_kg"], data["height_cm"])
    risk = 18.0
    family = data["family_history"]

    if bmi >= 25:
        risk += (bmi - 25) * 1.4
    if bmi >= 30:
        risk += 4
    if "Heart disease" in family:
        risk += 7
    if "Type 2 Diabetes" in family:
        risk += 6
    if "Hypertension" in family:
        risk += 5
    if data["sugar_intake"] == "High":
        risk += 6
    if data["sugar_intake"] == "Moderate":
        risk += 3
    if data["daily_steps"] < 7000:
        risk += (7000 - data["daily_steps"]) / 900
    if data["sleep_hours"] < 7:
        risk += (7 - data["sleep_hours"]) * 2.2
    if data["water_intake_liters"] < 2:
        risk += (2 - data["water_intake_liters"]) * 2
    if data["exercise_minutes"] < 30:
        risk += (30 - data["exercise_minutes"]) / 10
    if data.get("smoking") == "Current":
        risk += 10
    if data["systolic_bp"] >= 130:
        risk += 5

    risk -= clamp((data["daily_steps"] - 8000) / 1200, 0, 5)
    if 7 <= data["sleep_hours"] <= 8.5:
        risk -= 3
    if data["exercise_minutes"] >= 30:
        risk -= 3

    return int(round(clamp(risk, 6, 72)))


def calculate_domain_risks(profile: Any) -> dict[str, int]:
    data = profile_to_dict(profile)
    overall = calculate_overall_risk(data)
    bmi = calc_bmi(data["weight_kg"], data["height_cm"])
    family = data["family_history"]

    heart = clamp(
        overall * 0.85
        + (4 if bmi > 25 else 0)
        + (8 if "Heart disease" in family else 0)
        - (4 if data["daily_steps"] > 8000 else 0),
        5,
        80,
    )
    diabetes = clamp(
        overall * 0.9
        + (8 if bmi > 27 else 2)
        + (10 if "Type 2 Diabetes" in family else 0)
        + (8 if data["sugar_intake"] == "High" else 0),
        8,
        85,
    )
    hypertension = clamp(
        overall * 0.8
        + (10 if data["systolic_bp"] >= 130 else 0)
        + (7 if "Hypertension" in family else 0),
        6,
        80,
    )
    lifestyle = clamp(
        12
        + (10 if data["daily_steps"] < 6000 else 0)
        + (8 if data["sleep_hours"] < 6.5 else 0)
        + (9 if data["sugar_intake"] == "High" else 0)
        + (5 if data["water_intake_liters"] < 1.5 else 0)
        - (6 if data["exercise_minutes"] >= 30 else 0),
        5,
        70,
    )
    return {
        "overall": overall,
        "cardiovascular": int(round(heart)),
        "diabetes": int(round(diabetes)),
        "hypertension": int(round(hypertension)),
        "lifestyle": int(round(lifestyle)),
        "health_score": calculate_health_score(data),
        "risk_level": risk_level_from_percent(overall),
    }


def get_risk_contributions(profile: Any) -> list[dict[str, Any]]:
    data = profile_to_dict(profile)
    bmi = calc_bmi(data["weight_kg"], data["height_cm"])
    items: list[dict[str, Any]] = []

    bmi_impact = round((bmi - 24) * 3.2) if bmi >= 25 else round((24.5 - bmi) * -1.5)
    items.append({"key": "bmi", "factor": "BMI", "impact": int(bmi_impact), "detail": f"{bmi} kg/m²"})

    family = data["family_history"]
    family_impact = len(family) * 7 + 8 if family else -2
    items.append(
        {
            "key": "family",
            "factor": "Family history",
            "impact": int(family_impact),
            "detail": ", ".join(family) if family else "None reported",
        }
    )

    sugar = data["sugar_intake"]
    sugar_impact = 11 if sugar == "High" else 5 if sugar == "Moderate" else -4
    items.append({"key": "sugar", "factor": "Sugar intake", "impact": sugar_impact, "detail": sugar})

    sleep_impact = -8 if data["sleep_hours"] >= 7 else round((7 - data["sleep_hours"]) * 4)
    items.append(
        {"key": "sleep", "factor": "Sleep", "impact": int(sleep_impact), "detail": f"{data['sleep_hours']} hrs"}
    )

    activity_impact = -12 if data["daily_steps"] >= 8000 else round((8000 - data["daily_steps"]) / 450)
    items.append(
        {
            "key": "activity",
            "factor": "Regular activity",
            "impact": int(activity_impact),
            "detail": f"{int(data['daily_steps']):,} steps",
        }
    )

    water_impact = -5 if data["water_intake_liters"] >= 2 else round((2 - data["water_intake_liters"]) * 4)
    items.append(
        {
            "key": "water",
            "factor": "Hydration",
            "impact": int(water_impact),
            "detail": f"{data['water_intake_liters']} L",
        }
    )
    return sorted(items, key=lambda item: item["impact"], reverse=True)

File: src/utils/healthCalculations.js
export function calcBmi(weightKg, heightCm) {
  const h = Number(heightCm) / 100
  if (!h || !weightKg) return 0
  return Number((Number(weightKg) / (h * h)).toFixed(1))
}

export function bmiCategory(bmi) {
  if (bmi < 18.5) return { label: 'Underweight', tone: 'warn' }
  if (bmi < 25) return { label: 'Healthy', tone: 'good' }
  if (bmi < 30) return { label: 'Overweight', tone: 'warn' }
  return { label: 'High', tone: 'alert' }
}

export function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value))
}

export function riskStatus(percent) {
  if (percent < 20) return { label: 'Low Risk', tone: 'good' }
  if (percent < 40) return { label: 'Moderate Risk', tone: 'warn' }
  return { label: 'High Risk', tone: 'alert' }
}

export function scoreStatus(score) {
  if (score >= 80) return { label: 'Excellent', tone: 'good' }
  if (score >= 65) return { label: 'Good', tone: 'good' }
  if (score >= 50) return { label: 'Fair', tone: 'warn' }
  return { label: 'Needs attention', tone: 'alert' }
}

/**
 * Illustrative prototype scoring — not a medical model.
 */
export function calculateHealthScore(profile) {
  const bmi = calcBmi(profile.weight, profile.height)
  let score = 78

  if (bmi >= 18.5 && bmi < 25) score += 6
  else if (bmi >= 25 && bmi < 30) score -= 6
  else if (bmi >= 30) score -= 12
  else score -= 4

  score += clamp((profile.steps - 6000) / 400, -8, 8)
  score += clamp((profile.sleep - 7) * 4, -8, 6)
  score += clamp((profile.exerciseMinutes - 20) / 8, -6, 6)
  score += clamp((profile.water - 2) * 3, -5, 4)

  if (profile.sugar === 'High') score -= 7
  else if (profile.sugar === 'Moderate') score -= 3
  else score += 2

  if (profile.smoking === 'Current') score -= 10
  else if (profile.smoking === 'Former') score -= 3

  if (profile.familyHistory?.length) score -= profile.familyHistory.length * 2

  return Math.round(clamp(score, 28, 96))
}

export function calculateOverallRisk(profile) {
  const bmi = calcBmi(profile.weight, profile.height)
  let risk = 18

  if (bmi >= 25) risk += (bmi - 25) * 1.4
  if (bmi >= 30) risk += 4
  if (profile.familyHistory?.includes('Heart disease')) risk += 7
  if (profile.familyHistory?.includes('Type 2 Diabetes')) risk += 6
  if (profile.familyHistory?.includes('Hypertension')) risk += 5
  if (profile.sugar === 'High') risk += 6
  if (profile.sugar === 'Moderate') risk += 3
  if (profile.steps < 7000) risk += (7000 - profile.steps) / 900
  if (profile.sleep < 7) risk += (7 - profile.sleep) * 2.2
  if (profile.water < 2) risk += (2 - profile.water) * 2
  if (profile.exerciseMinutes < 30) risk += (30 - profile.exerciseMinutes) / 10
  if (profile.smoking === 'Current') risk += 10
  if (profile.systolic >= 130) risk += 5

  risk -= clamp((profile.steps - 8000) / 1200, 0, 5)
  if (profile.sleep >= 7 && profile.sleep <= 8.5) risk -= 3
  if (profile.exerciseMinutes >= 30) risk -= 3

  return Number(clamp(risk, 6, 72).toFixed(0))
}

export function calculateDomainRisks(profile) {
  const overall = calculateOverallRisk(profile)
  const bmi = calcBmi(profile.weight, profile.height)

  const heart = clamp(
    overall * 0.85 + (bmi > 25 ? 4 : 0) + (profile.familyHistory?.includes('Heart disease') ? 8 : 0) - (profile.steps > 8000 ? 4 : 0),
    5,
    80,
  )
  const diabetes = clamp(
    overall * 0.9 + (bmi > 27 ? 8 : 2) + (profile.familyHistory?.includes('Type 2 Diabetes') ? 10 : 0) + (profile.sugar === 'High' ? 8 : 0),
    8,
    85,
  )
  const hypertension = clamp(
    overall * 0.8 + (profile.systolic >= 130 ? 10 : 0) + (profile.familyHistory?.includes('Hypertension') ? 7 : 0),
    6,
    80,
  )
  const lifestyle = clamp(
    12 +
      (profile.steps < 6000 ? 10 : 0) +
      (profile.sleep < 6.5 ? 8 : 0) +
      (profile.sugar === 'High' ? 9 : 0) +
      (profile.water < 1.5 ? 5 : 0) -
      (profile.exerciseMinutes >= 30 ? 6 : 0),
    5,
    70,
  )

  return {
    overall,
    heart: Math.round(heart),
    diabetes: Math.round(diabetes),
    hypertension: Math.round(hypertension),
    lifestyle: Math.round(lifestyle),
  }
}

export function getRiskContributions(profile) {
  const bmi = calcBmi(profile.weight, profile.height)
  const items = []

  const bmiImpact = bmi >= 25 ? Math.round((bmi - 24) * 3.2) : Math.round((24.5 - bmi) * -1.5)
  items.push({ key: 'bmi', label: 'BMI', value: bmiImpact, detail: `${bmi} kg/m²` })

  const family = profile.familyHistory?.length ? profile.familyHistory.length * 7 + 8 : -2
  items.push({ key: 'family', label: 'Family history', value: family, detail: profile.familyHistory?.length ? profile.familyHistory.join(', ') : 'None reported' })

  const sugar = profile.sugar === 'High' ? 11 : profile.sugar === 'Moderate' ? 5 : -4
  items.push({ key: 'sugar', label: 'Sugar intake', value: sugar, detail: profile.sugar })

  const sleep = profile.sleep >= 7 ? -8 : Math.round((7 - profile.sleep) * 4)
  items.push({ key: 'sleep', label: 'Sleep', value: sleep, detail: `${profile.sleep} hrs` })

  const activity = profile.steps >= 8000 ? -12 : Math.round((8000 - profile.steps) / 450)
  items.push({ key: 'activity', label: 'Activity', value: activity, detail: `${profile.steps.toLocaleString()} steps` })

  const water = profile.water >= 2 ? -5 : Math.round((2 - profile.water) * 4)
  items.push({ key: 'water', label: 'Hydration', value: water, detail: `${profile.water} L` })

  return items.sort((a, b) => b.value - a.value)
}

export function simulateRisk(baselineProfile, scenario) {
  const merged = { ...baselineProfile, ...scenario }
  return calculateDomainRisks(merged).overall
}

export function percentChange(current, previous) {
  if (!previous) return 0
  return Number((((current - previous) / previous) * 100).toFixed(1))
}
